Firestore/core/src/util/executor_libdispatch.mm

```diff
@@ -243,91 +94,231 @@ void MarkDone() {
   namespace chr = std::chrono;
   const dispatch_time_t delay_ns = dispatch_time(
       DISPATCH_TIME_NOW, chr::duration_cast<chr::nanoseconds>(delay).count());
 
-  // Ownership is fully transferred to libdispatch -- because it's impossible
-  // to truly cancel work after it's been dispatched, libdispatch is
-  // guaranteed to outlive the executor, and it's possible for work to be
-  // invoked by libdispatch after the executor is destroyed. Executor only
-  // stores an observer pointer to the operation.
-  TimeSlot* time_slot = nullptr;
-  TimeSlotId time_slot_id = 0;
-  RunSynchronized(this, [this, delay, &operation, &time_slot, &time_slot_id] {
-    time_slot_id = NextId();
-    time_slot = new TimeSlot{this, delay, std::move(operation), time_slot_id};
-    schedule_[time_slot_id] = time_slot;
-  });
-
-  dispatch_after_f(delay_ns, dispatch_queue(), time_slot,
-                   TimeSlot::InvokedByLibdispatch);
-
-  return DelayedOperation{[this, time_slot_id] {
-    // `time_slot` might have been destroyed by the time cancellation function
-    // runs, in which case it's guaranteed to have been removed from the
-    // `schedule_`. If the `time_slot_id` refers to a slot that has been
-    // removed, the call to `RemoveFromSchedule` will be a no-op.
-    RemoveFromSchedule(time_slot_id);
-  }};
+  // Ownership is shared with libdispatch because it's impossible to actually
+  // cancel work after it's been dispatched, libdispatch is guaranteed to
+  // outlive the executor, and it's possible for work to be invoked by
+  // libdispatch after the executor is destroyed. While the Executor has a
+  // pointer to the Task it also has ownership.
+  Task* task = nullptr;
+  TimePoint target_time = MakeTargetTime(delay);
+  Id id = 0;
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    id = NextIdLocked();
+    task = Task::Create(this, target_time, tag, id, std::move(operation));
+    task->Retain();  // For libdispatch's ownership
+
+    tasks_.insert(task);
+    schedule_[id] = task;
+  }
+
+  dispatch_after_f(delay_ns, dispatch_queue_, task, InvokeAsync);
+
+  return DelayedOperation(this, id);
 }
 
-void ExecutorLibdispatch::RemoveFromSchedule(TimeSlotId to_remove) {
-  RunSynchronized(this, [this, to_remove] {
-    const auto found = schedule_.find(to_remove);
+void ExecutorLibdispatch::OnCompletion(Task* task) {
+  bool should_release = false;
+  {
+    TASK_TRACE("Executor::OnCompletion %s task %s", this, task);
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto found = tasks_.find(task);
+    if (found != tasks_.end()) {
+      should_release = true;
+      tasks_.erase(found);
+
+      // Only try to remove scheduled tasks here because non-scheduled tasks
+      // all have id 0, which overlaps with the first scheduled task.
+      if (!task->is_immediate()) {
+        schedule_.erase(task->id());
+      }
+    }
+  }
+
+  if (should_release) {
+    task->Release();
```

Comment:
In the presence of calls between objects that each have their own locks, a general deadlock avoidance protocol is to avoid making calls to others while holding your own lock.

`Release` does not acquire the `Task`'s lock and does not call back to the executor so I guess it's safe to call while holding the executor's lock, but I'm wary of abandoning the general rule on the basis of this knowledge. Task could change in such a way that it called back into executor, or even worse could change such that it blocked on some other thread's callback into the executor and then it would introduce the possibility of deadlock (the former case can be solved by using recursive locks, the latter cannot).

I've added a comment here stating:

> Avoid calling potentially locking methods on the task while holding the executor's lock.